lib/removeUnusedVertices.js

```diff
@@ -0,0 +1,364 @@
+    for (var meshId in meshes) {
+        if (meshes.hasOwnProperty(meshId)) {
+            if (defined(meshes[meshId].primitives)) {
+                primitives.push.apply(primitives, meshes[meshId].primitives);
+            }
+        }
+    }
+    calculateIndexIntervals(accessors, bufferViews, buffers, primitives);
+
+    compressBuffers(accessors, bufferViews, buffers);
+
+    //Recalculate the byte length for each buffer.
+    for (var bufferId in buffers) {
+        if (buffers.hasOwnProperty(bufferId)) {
+            var buffer = buffers[bufferId];
+            if (defined(buffer.extras._pipeline.offset)) {
+                buffer.byteLength = buffer.extras._pipeline.source.length;
+            }
+        }
+    }
+    
+    updateIndices(bufferViews, buffers);
+}
+
+//Calculate the range of indices used by primitives and their accessors.
+function calculateIndexIntervals(accessors, bufferViews, buffers, primitives) {
+    for (var bufferViewId in bufferViews) {
+        if (bufferViews.hasOwnProperty(bufferViewId)) {
+            bufferViews[bufferViewId].extras._pipeline.accessors = {};
+        }
+    }
+
+    for (var i = 0; i < primitives.length; i++) {
+        var primitive = primitives[i];
+        if (defined(primitive.indices)) {
+            var indexAccessor = accessors[primitive.indices];
+            var indexView = bufferViews[indexAccessor.bufferView];
+            var indexBuffer = buffers[indexView.buffer];
+            var indexSource = indexBuffer.extras._pipeline.source.slice(indexView.byteOffset, indexView.byteOffset + defaultValue(indexView.byteLength, 0));
+            var indexStride = indexAccessor.componentType === WebGLConstants.UNSIGNED_SHORT ? 2 : 1;
+            indexView.extras._pipeline.indexStride = defaultValue(indexView.extras._pipeline.indexStride, indexStride);
+            indexView.extras._pipeline.indexIntervals = defaultValue(indexView.extras._pipeline.indexIntervals, []);
+
+            //Create a sorted array of unique index values.
+            var indexArray = [];
+            var accessorSource = indexSource.slice(indexAccessor.byteOffset, indexAccessor.byteOffset + indexStride * indexAccessor.count);
+            var accessorIndex = 0;
+            while (accessorIndex < accessorSource.length) {
+                var index = WebGLConstants.UNSIGNED_SHORT ? accessorSource.readUInt16LE(accessorIndex) : accessorSource.readUInt8LE(accessorIndex);
+                if (indexArray.indexOf(index) === -1) {
+                    indexArray.push(index);
+                }
+                accessorIndex += indexStride;
+            }
+            indexArray.sort(function(a, b) {
+                return a - b;
+            });
+
+            //Calculate the index intervals based on the array of sorted indices.
+            var indexIntervals = [];
+            var intervalStart, prevIndex;
+            var newAccessorCount = 0;
+
+            for (var j = 0; j < indexArray.length; j++) {

```

Comment:
Throughout this file, we tend to write loops like this with (unless the length potential changes):
```javascript
var length = indexArray.length;
for (var j = 0; j < length; j++) {
```
to avoid the potential lookup each time.